File: 23. Merge k Sorted Lists.cpp
// Time:  O(n)
// Space: O(1)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0)
            return NULL;
        while (lists.size() > 1) { // merge first two at a time
            lists.push_back(mergeTwoLists(lists[0], lists[1]));
            lists.erase(lists.begin());
            lists.erase(lists.begin());
        }
        return lists.front();
    }
    
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode preHead(INT_MIN);
        ListNode *p = &preHead;
        while (l1 && l2) {
            if (l1->val <= l2->val) { // smaller one rank first
                p->next = l1;
                l1 = l1->next;
            } else {
                p->next = l2;
                l2 = l2->next;
            }
            p = p->next;
        }
        p->next = l1 ? l1 : l2;
        return preHead.next;
    }
};

File: 23. Merge k Sorted Lists.py
# 23. Merge k Sorted Lists
# Time:  O(Nlogk)
# Space: O(k)

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# use heap to maintain the streaming max/min value
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        heap = []
        for l in lists:
            if l:
                heapq.heappush(heap, (l.val, l))
        
        cur = dummy = ListNode(0)
        while heap:
            min_l = heapq.heappop(heap)[1]
            cur.next = min_l
            cur = cur.next
            if min_l.next:
                heapq.heappush(heap, (min_l.next.val, min_l.next))
        return dummy.next
